_posts/2024-01-10-semantics-of-regular-expressions.markdown

```diff
@@ -0,0 +1,386 @@
+---
+layout: post
+title:  "Well-Behaved (Co)algebraic Semantics of Regular Expressions in Dafny"
+date:   2024-01-10 18:00:00 +0100
+author: Stefan Zetzsche, Wojciech Różowski
+---
+
+## Introduction
+
+[Regular expressions](https://en.wikipedia.org/wiki/Regular_expression) are one of the most ubiquitous formalisms of theoretical computer science. Commonly, they are understood in terms of their [*denotational* semantics](https://en.wikipedia.org/wiki/Denotational_semantics), that is, through the formal languages — the [*regular* languages](https://en.wikipedia.org/wiki/Regular_language) — that they induce. This view is *inductive* in nature: two primitives are equivalent if they are *constructed* in the same way.  Alternatively, regular expressions can be understood in terms of their [*operational* semantics](https://en.wikipedia.org/wiki/Operational_semantics), that is, through the [finite automata](https://en.wikipedia.org/wiki/Finite-state_machine) they induce. This view is *coinductive* in nature: two primitives are equivalent if they are *deconstructed* in the same way. It is is hinted at by [Kleene’s famous theorem](http://www.dlsi.ua.es/~mlf/nnafmc/papers/kleene56representation.pdf) that both views are equivalent: regular languages are precisely the formal languages accepted by finite automata. In this blogpost, we utilise Dafny’s built-in inductive and coinductive reasoning capabilities to show that the two semantics of regular expressions are *[well-behaved](https://homepages.inf.ed.ac.uk/gdp/publications/Math_Op_Sem.pdf)*, in the sense they are in fact one and the same, up to pointwise [bisimulation](https://en.wikipedia.org/wiki/Bisimulation). 
+
+## Denotational Semantics
+
+In this section, we define regular expressions and formal languages, introduce the concept of bisimilarity, formalise the *denotational* semantics of regular expressions as a function from regular expressions to formal languages, and prove that the latter is an algebra homomorphism.
+
+### Regular Expressions as Datatype
+
+We define the set of regular expressions parametric in an alphabet  `A` as an inductive [`datatype`](https://dafny.org/latest/DafnyRef/DafnyRef#sec-inductive-datatypes):
+
+```
+datatype Exp<A> = Zero | One | Char(A) | Plus(Exp, Exp) | Comp(Exp, Exp) | Star(Exp)
+```
+
+The definition captures that a regular expression is either a primitive character `Char(a)`,  a non-deterministic choice between two regular expressions `Plus(e1, e2)`, a sequential iteration of two regular expressions `Comp(e1, e2)`,  a finite number of self-iterations `Star(e)`, or one of the constants `Zero` (the unit of `Plus`) and `One` (the unit of `Comp`). On a more high-level, above defines  `Exp<A>` as the *smallest* algebraic structures that is equipped with two constants, contains all elements of type `A`, and is closed under two binary and one unary operation.
```

Comment:
I'm not sure what "sequential iteration" is supposed to mean. Why not just call it "concatenation"? Similarly, can we say something like "Zero" is the empty set and "One" the empty string, just to give more intuition to the readers?

Reply:
As suggested by Aaron, I'll go with sequential _composition_.